### ISSUE TITLE
Error get_body_agrs()

### ISSUE THREAD

**Issue by nickwq** [Kong/kong#3055]
### Summary

From kong log file error.log, we got lots of error like this:

------------------------------------------------------------

2017/11/29 08:35:59 [error] 51#0: *1793068 [lua] public.lua:27: get_body_args(): could not decode JSON body args: bad argument #1 to '?' (string expected, got nil), client: 10.2.50.0, server: kong, request: "GET /s/users HTTP/1.0", host: "gate.test.com", referrer: "https://test.com/wx7c861572cd0550c0/15/page-frame.html"

----------------------------------------------------------------------

So what does body args mean? A GET request actually has no body data. Why this error occurs?
Your feedback and suggestions are appreciated, thanks!

### Additional Details & Logs

- Kong version (0.11)
- Operating System (COREOS)

**Comment by WALL-E** [comment]
Which plugins are used for the API(/s/users)?

**Comment by nickwq** [comment]
@WALL-E Thanks for replying. I have two global plugins 'oatuh2' and 'cors'. Some of the APIs have kay-auth, but not this one. So you mean this issue could be related to plugins?

**Comment by WALL-E** [comment]
It seems that this error message can be ignored.

## version 0.11.0
```
    [2] = function() -- json
      local body, err = cjson.decode(ngx_req_get_body_data())
      if err then
        ngx_log(ERR, "could not decode JSON body args: ", err)
        return {}
      end

      return body
    end,
```
It is necessary to check whether the return value of the function((ngx_req_get_body_data) is nil 

## version 0.11.2
```
    [MIME_TYPES.json] = function()
      local raw_body  = ngx_req_get_body_data()
      local args, err = cjson.decode(raw_body)
      if err then
        ngx_log(ERR, "could not decode JSON body args: ", err)
        return {}, raw_body
      end

      return args, raw_body
    end,
```
v11.2 has the same problem, I'll fix it.

**Comment by WALL-E** [comment]
this issue was related to the plugin oauth2

**Comment by nickwq** [comment]
@WALL-E you mean is a bug of OAUTH2 or incorrect usage of oauth2?

**Comment by WALL-E** [comment]
[plugins/oauth2/access.lua#L94,L99](https://github.com/Kong/kong/blob/0.11.0/kong/plugins/oauth2/access.lua#L94,L99)

**Comment by nickwq** [comment]
We pass the token by setting the header  like this:  
"Authorization": "Bearer JiJW4jNuKXyyRotTQw12381CpXsSb"

I think the code logic has a bug,  request like GET actually has no body, better check whether the body is null before actually parse it.

**Comment by WALL-E** [comment]
PR #3059

**Comment by thibaultcha** [comment]
Hi!

Indeed, there is definitely something wrong going on here... It is my belief that if you were to retrieve the full stacktrace to that error @nickwq (that would be informative), it would show that this originated from [this call](https://github.com/Kong/kong/blob/0.11.0/kong/plugins/oauth2/access.lua#L406), which isn't running in `POST` request only, unlike other calls to `retrieve_parameters()` in this module.

@WALL-E Thank you for the patch.

Ultimately, I believe this is an issue of the OAuth2 plugin itself: there should be no reason to call `ngx.req.read_body()` or `public_utils.get_body_args()` for requests that do not have a payload. The later should probably log some error when that is being attempted, so a fix to the OAuth2 plugin would still be in order. Maybe a safeguard inside of this `retrieve_parameters()` function ensuring that the request indeed has a payload? That'd be a safer approach.

**Comment by nickwq** [comment]
@thibaultcha thanks for your explanation.

**Comment by thibaultcha** [comment]
This should be fixed by #3063, thank you for the report!
